### ISSUE TITLE
[question] Generating CMake config files AFTER deployment

### ISSUE THREAD

**Issue by NicolasIRAGNE** [conan-io/conan#9874]
Forgive me if the question has been submitted already, I have been googling a bunch but no avail (have learned a lot about Conan the Barbarian though, thanks google). 

TL;DR at the end if you just want the question

So: here's my situation. I am currently working on a project with a team, and that project has a few annoying dependencies (I'm talking CUDA, OpenCV, TensorRT and the like).

My initial idea was providing some sort of "SDK": an archive of all the dependencies that are pre-built and **ready to be consumed by CMake without changing anything.** Currently we have to build / install every dependency individually and we use Find_Package() to link them.

So my idea was having an automated system (very likely to be a gitlab-runner) hosted on our server that would provide standard "packages" (read: just plain .tar archives with binaries/headers/libs). Then I would just have to download and extract everything somewhere and set CMAKE_PREFIX_PATH to that directory when compiling. Easy enough, right?
So far I have managed to pretty much do that: I am properly building and packaging everything. However, I still have one major problem: **The files generated by the cmake generators point to the local cache.** I understand this is the intended behavior, but in my case this renders everything unusable.

Here is what I have done so far:
  * First, I created recipes for most of my dependencies, as I was not satisfied with the ones on conan center (and mostly because I had to do some unorthodox things like installing CUDA). I specify this because then I can add anything I need to the recipes if needed.
  * I export all my recipes in the local cache (so on the machine that will make the packages available to download)
  * Then I create a conanfile for the sdk. This file is simply used to specify which packages I want to deploy. At first I used a .txt file that just stated the dependencies and the imports folders. That worked great! The only problem is that all the cmakeConfig files that are generated are not setting the path correctly.
  * I also tried the deploy generator, which seems to work for libraries that already have a CMake config. However, ``nlohmann_json`` for example, does not ship one but instead creates it upon calling cmake install.

I am a bit short of ideas now. I have read a bit about the generate() and deploy() methods but cannot really figure out what to do with them.

### So, TL;DR:
I have a conanfile that simply imports() all its dependencies. I want to make it so they are usable **as is**, without conan, so they can be used anywhere. The cmake generator seems to work well but **it points to my local cache instead of having something relative** (I'm not sure something like this is possible -- but maybe it can be generated after the deploy?). Is there a way to tell the cmake generator that the files are in a different location? Having the user specify an extra command upon installing is fine I guess.

Is there a way to do this within Conan? Or do I have to implement it externally? (for example patching the existing cmake files with a script). Do I have to write my own CMake config files? I feel like I am so close to making it working and I'm just missing something and it is getting frustrating.

Note: This has to be done just for deployment. Some of my packages are dependent on each other and ideall I would like a way to not break them.

I am quite a beginner in terms of software development and I am... lost. Sorry for the confusion.

Thank you :-)

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by NicolasIRAGNE** [comment]
For the record, this is what my conanfile.txt looks like. The packages with @py/local are my own. 
I liked the conanfile.txt approach because it made editing the file a lot easier for my coworkers without them having to learn much about conan, but I switched to a .py file because I'm assuming it allows for more flexibility.

```
[requires]
nlohmann_json/3.10.4@py/local
opencv/4.4.0@py/local
openssl/3.0.0
boost/1.77.0
tensorrt/7.2.3.4@py/local
ffmpeg/4.4@py/local

[generators]
cmake_find_package_multi
cmake_paths

[options]
boost:shared=True
ffmpeg:with_gpu=True
ffmpeg:shared=True
ffmpeg:with_programs=True
opencv:shared=True
opencv:contrib=True
opencv:contrib_freetype=False
opencv:contrib_sfm=False
opencv:parallel=tbb
opencv:with_ade=False
opencv:with_jpeg=libjpeg
opencv:with_png=True
opencv:with_tiff=False
opencv:with_jpeg2000=False
opencv:with_openexr=False
opencv:with_eigen=True
opencv:with_webp=False
opencv:with_quirc=False
opencv:with_cuda=True
opencv:with_cublas=True
opencv:with_cufft=False
opencv:with_imgcodec_hdr=False
opencv:with_imgcodec_pfm=False
opencv:with_imgcodec_pxm=False
opencv:with_imgcodec_sunraster=False
opencv:dnn=True
opencv:detect_cpu_baseline=False
opencv:nonfree=True
opencv:with_ffmpeg=True

[imports]
bin, * -> ./bin
lib, * -> ./lib
include, * -> ./include @ excludes=*opencv4*
include/opencv4, * -> ./include
```

**Comment by memsharded** [comment]
Hi @NicolasIRAGNE 

As a first hint, I would probably recommend to start trying the new ``CMakeDeps`` generator, instead of the ones you are using, read in this section: https://docs.conan.io/en/latest/reference/conanfile/tools.html. 
The main reason is because the generated ``xxxxxxx-data.cmake`` files written by ``CMakeDeps`` have already extracted a common cmake variable, and reused it along the file. So if going to try patching the generated files, this could be easier and cleaner. (Note, ``CMakeDeps`` plays with the new ``CMakeToolchain`` and new ``CMake`` helper from ``from conan.tools.cmake import xxx``).

Then, yes, I'd probably use the ``generate()`` method to change that, something in the line:

```python
def generate(self):
     deps = CMakeDeps(self):
     files = deps.content  # instead of calling ``generate()``, I get the files and patch them before saving
     for generator_file, content in files.items():
         if generator_file.endswith("data.cmake"):
                # line to replace: set({{ pkg_name }}_PACKAGE_FOLDER{{ config_suffix }} "{{ package_folder }}")
                # use a regex from _PACKAGE_FOLDER[_RELEASE|_DEBUG...] "(group)" ) to extract the "group" that is the folder
                content = content.replace(<original-pkg-folder>, "my/desired/location")
         save(generator_file, content)
```

I know it seems a bit tricky, but should be possible, we have designed the new integrations so these things are way more flexible. We also know that this is not ideal, and we have considered some different ways to try to improve it, but so far it has been quite challenging to come up with something generic. The main problem is that the final layout defined by the ``impors`` has no guarantee of being usable or that it avoids binary collisions. We will most likely keep investigating on this soon, but as it seems you need this now, I'd certainly try something as described above. Please let me know if you have any question.

**Comment by memsharded** [comment]
You might need to move your ``conanfile.txt`` to a ``conanfile.py``, but using declarative syntax like ``requires = "req1..."``, the difference will be little, still very readable and simple conanfile

**Comment by NicolasIRAGNE** [comment]
Thanks a lot. The CMakeDeps indeed looks more like what I want. I will investigate tomorrow (or maybe tonight because this is bugging me). I will keep you posted. 

(also I already have converted my .txt to .py, I just posted this one because it was more readable on here and I have not implemented anything new at the moment)

Sorry for the rush. It's just that the sooner I can get something working, the sooner I can relieve my coworkers of all the hassle of building so they can focus on the project itself. I should take a bit more time to read the manual more intensively.

Thank you.

**Comment by NicolasIRAGNE** [comment]
So, if I understand correctly, this means that the cmake files will be patched when all the packages are imported from the main conanfile yes ? 

So if I move the SDK the files will be invalid and will need to be patched again, yes? 
I would be fine with this but if my understanding is correct then using this bit of code requires conan installed (or should be easy enough to rewrite in vanilla python i guess).

I am not proficient enough in CMake to really understand the best way to do it. 
I'm guessing it's not possible to have relocatable CMake Config files (as in relative paths)? This might be a bit more complicated than I thought 😅

Sorry if these questions are naive, I am not very proficient in Python.

**Comment by memsharded** [comment]
Probably not necessary. You could define for example one generic ``MY_DEPS_FOLDER`` cmake variable, and patch it in the generated files from ``CMakeDeps``, not patch a fixed, absolute path. Then assume the consumer ``CMakeLists.txt`` must define this value to point to the SDK location, **before** calling any ``find_package()``.

**Comment by NicolasIRAGNE** [comment]
Thank you. I think I see what you mean. This seems like the way to go, I will try to implement it.

There's one thing I am not sure: in the main conanfile generate() method, how do I get the original package path in order to patch it ? 

Now that I think about it, this is probably what the regex you gave me is for, though I am not sure how to implement it :(
 
Thanks **a lot** for your time. I really appreciate it. You're saving my buttcheeks here.

Edit: after fiddling a bit with regex, I got this: ```(?:set\([A-Za-z]+_PACKAGE_FOLDER(?:_RELEASE|_DEBUG)).+"(.*?)"``` which kinda works, now I'm trying to replace the part in quotes from python :)

**Comment by memsharded** [comment]
> There's one thing I am not sure: in the main conanfile generate() method, how do I get the original package path in order to patch it ?

> Now that I think about it, this is probably what the regex you gave me is for, though I am not sure how to implement it :(

Yes, exactly, that is the idea to replace that path. It seems you are in the right direction with your regex (if you are using only Release and Debug builds, and not RelWithDebInfo). Recall that you don't need to do the replace yourself, you can use ``re.sub()`` to do the replacement for you. The key is getting the group to replace correctly. In case you don't know it, I love https://regex101.com

**Comment by NicolasIRAGNE** [comment]
Alriiiiight so my generate() method looks like this now
```python
def generate(self):
        deps = CMakeDeps(self)
        files = deps.content
        print("Install folder:" + self.install_folder)
        for generator_file, content in files.items():
         if generator_file.endswith("data.cmake"):
                print("patching " + generator_file + "...")
                match = re.search(r'(?:set\([a-zA-Z0-9_-]+_PACKAGE_FOLDER(?:_RELEASE|_DEBUG)).+"(.*?)"', content)
                if (match):
                    patched_string = re.sub(r"\"(.*?)\"", "\"${SDK_DIR}\"", match.group(0))
                    new_content = content.replace(match.group(0), patched_string)
                    save(generator_file, new_content)
                    print("successfully patched " + generator_file)
                else:
                    print("failed to patch " + generator_file + ": no match found")
```

It seems to be working. I managed to link against some packages (like nlohmann_json and boost). For now I can't link against some of my packages because of things like:
```
CMake Error at /tmp/full/cmakedeps_macros.cmake:4 (message):
  Library 'opencv_dnn' not found in package.  If 'opencv_dnn' is a system
  library, declare it with 'cpp_info.system_libs' property
Call Stack (most recent call first):
  /tmp/full/cmakedeps_macros.cmake:48 (conan_message)
  /tmp/full/OpenCV-Target-release.cmake:21 (conan_package_library_targets)
  /tmp/full/OpenCVTargets.cmake:27 (include)
  /tmp/full/OpenCVConfig.cmake:11 (include)
  CMakeLists.txt:40 (find_package)
```

but it's very likely that this is coming from my OpenCV recipe, so I will investigate that tomorrow because right now i have no clue what this means and I'm exhausted. I deserve a raise.

Thanks a lot for your help. I can't stress this enough. I really appreciate it.
I'll keep you posted tomorrow.

**Comment by NicolasIRAGNE** [comment]
Hello again,

I got it working :partying_face: :tada: .

I am so relieved. I had to tweak some packages here and there for them to generate correct targets for CMake but otherwise i managed to compile my project with it. I'm still having some issues with symbolic links here and there but it's just small typos.

Many thanks to you.

**Comment by memsharded** [comment]
🎉 🎉 🎉 🎉 😄 Great, happy to hear that!

I guess that you have managed to make it work making sure the libs are correctly packaged, that the ``set_property`` for custom filenames and custom target names (in case you needed them), and tuning the ``package_info()`` of the recipes in general.

Thanks to you for following up. If this is solved, I think we can close it, but I would suggest to keep en eye on the releases (there is an email list for releases) and changelogs, because at some point we will try to make something more built-in for this.

**Comment by NicolasIRAGNE** [comment]
Indeed, I had some trouble with some case sensitive target but nothing a quick package_info() tweak can't fix.

I will keep an eye on releases, this is definitely a tool I find very useful. 
I hesitated a long time between 1) using Conan for what I wanted to do even though it wasn't *exactly* what it was made for and 2) quickly developing a rudimentary tool instead. In the end, I am glad I chose Conan because it will be a lot more maintanable in the future. 

Now, on to windows compatibility!
